contracts/PodsUpkeep.sol

```diff
@@ -0,0 +1,159 @@
+    function _readLastBlockNumberForPodIndex(uint256 _existingUpkeepBlockNumbers, uint8 _podIndex) internal pure returns (uint32) { 
+  
+        uint256 mask =  (type(uint32).max | uint256(0)) << (_podIndex * 32);
+        return uint32((_existingUpkeepBlockNumbers & mask) >> (_podIndex * 32));
+    }
+
+    /// @notice Get the last Upkeep block number for a pod
+    /// @param podIndex The position of the pod in the Registry
+    function readLastBlockNumberForPodIndex(uint256 podIndex) public view returns (uint32) {
+        
+        uint256 wordIndex = podIndex / 8;
+        return _readLastBlockNumberForPodIndex(lastUpkeepBlockNumber[wordIndex], uint8(podIndex % 8));
+    }
+
+    /// @notice Checks if Pods require upkeep. Call in a static manner every block by the Chainlink Upkeep network.
+    /// @param checkData Not used in this implementation.
+    /// @return upkeepNeeded as true if performUpkeep() needs to be called, false otherwise. performData returned empty. 
+    function checkUpkeep(bytes calldata checkData) override external view returns (bool upkeepNeeded, bytes memory performData) {
+        
+        address[] memory pods = podsRegistry.getAddresses();
+        uint256 _upkeepBlockInterval = upkeepBlockInterval;
+        uint256 podsLenght = pods.length;
+
+        for(uint256 podWord = 0; podWord <= podsLenght / 8; podWord++){
+
+            uint256 _lastUpkeep = lastUpkeepBlockNumber[podWord]; // this performs the SLOAD
+            for(uint256 i = 0; i + (podWord * 8) < podsLenght; i++){
+                
+                uint32 podLastUpkeepBlockNumber = _readLastBlockNumberForPodIndex(_lastUpkeep, uint8(i));
+                if(block.number > podLastUpkeepBlockNumber + _upkeepBlockInterval){
+                    return (true, "");
+                }
+            }
+        }       
+        return (false, "");    
+    }
+
+    /// @notice Performs upkeep on the pods contract and updates lastUpkeepBlockNumbers
+    /// @param performData Not used in this implementation.
+    function performUpkeep(bytes calldata performData) override external {
+    
+        address[] memory pods = podsRegistry.getAddresses();
+        uint256 podsLenght = pods.length;
+        uint256 _batchLimit = upkeepBatchLimit;
+        uint256 batchesPerformed = 0;
+
+        for(uint8 podWord = 0; podWord <= podsLenght / 8; podWord++){ // give word index
+            
+            uint256 _updateBlockNumber = lastUpkeepBlockNumber[podWord]; // this performs the SLOAD
+
+            for(uint8 i = 0; i + (podWord * 8) < podsLenght; i++){ // pod index within word
+                
+                if(batchesPerformed >= _batchLimit) {
+                    break;
+                }
+                uint32 podLastUpkeepBlockNumber = _readLastBlockNumberForPodIndex(_updateBlockNumber, i);
+                // what happens when a pod is removed from the registry -- zero out the address and insert a callStatic check for .batch() here?
```

Comment:
@asselstine I think it may be easier to zero out in the Registry in the case of a removed pod?